x-pack/plugins/observability/public/pages/cases/helpers.ts

```diff
@@ -4,16 +4,98 @@
  * 2.0; you may not use this file except in compliance with the Elastic License
  * 2.0.
  */
-import { useEffect, useState } from 'react';
+import { useEffect, useState, useCallback } from 'react';
 import { isEmpty } from 'lodash';
+
+import type * as estypes from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
 import { usePluginContext } from '../../hooks/use_plugin_context';
 import { TopAlert, parseAlert } from '../../pages/alerts/';
 import { useKibana } from '../../utils/kibana_react';
-import { Ecs } from '../../../../cases/common';
+import { BASE_RAC_ALERTS_API_PATH } from '../../../../rule_registry/common/constants';
+
+export const useFetchAlertData = (alertIds: string[]): [boolean, Record<string, unknown>] => {
+  const { http } = useKibana().services;
+  const [loading, setLoading] = useState(false);
+  const [alerts, setAlerts] = useState<Record<string, unknown> | undefined>(undefined);
+
+  const fetch = useCallback(
+    async (ids: string[], abortCtrl: AbortController) => {
+      try {
+        setLoading(true);
+        const response = await http.post<estypes.SearchResponse<Record<string, unknown>>>(
+          `${BASE_RAC_ALERTS_API_PATH}/find`,
+          {
+            body: JSON.stringify({
+              query: {
+                ids: {
+                  values: ids,
+                },
+              },
+              track_total_hits: false,
+              size: 10000,
+            }),
+            signal: abortCtrl.signal,
+          }
+        );
+
+        if (response) {
+          console.log(JSON.stringify(getAlertsGroupedById(response), null, 2));
+          setAlerts(getAlertsGroupedById(response));
+        }
+      } catch (error) {
+        console.log('got an error ', error);
+        setAlerts(undefined);
+      } finally {
+        setLoading(false);
+      }
+    },
+    [http]
+  );
+
+  const validIds = getValidValues(alertIds);
+  console.log('valid ids', validIds);
+
+  useEffect(() => {
+    const abortController = new AbortController();
+    console.log('rerendering');
+    if (shouldFetch(validIds, alerts)) {
+      console.log('fetching');
+      fetch(validIds, abortController);
+    } else {
+      console.log('not fetching');
+    }
+
+    return () => {
+      // abortController.abort();
```

Comment:
@cnasikas @academo I'm doing something wrong here, as soon as I uncomment this it causes a rerender infinite loop because the previous request is like immediately getting canceled and the `alerts` field is always `undefined`.

I'm also unsure if it is rerendering because the `validIds` reference is changing. I tried memoizing that but it didn't help. From my research it seems like react's shallowEquals will return false for arrays that have the same values since their memory reference is different right?

Reply:
@michaelolo24 helped me figure it out :) we need to add some memoization for the `alertIds` being passed in.